nemo/backends/pytorch/actions.py

```diff
@@ -1093,6 +1010,179 @@ def train(
+                        )
+
+        def _init_callbacks(callbacks, action):
+            if callbacks is not None and isinstance(callbacks, List) and len(callbacks) > 0:
+                for callback in callbacks:
+                    if isinstance(callback, ActionCallback):
+                        callback.action = action
+
+        def _update_callbacks(callbacks=None, registered_tensors=None, final_loss=None):
+            # if self.local_rank is None or self.local_rank == 0:
+            if callbacks is not None and isinstance(callbacks, List) and len(callbacks) > 0:
+                for callback in callbacks:
+                    if isinstance(callback, ActionCallback):
+                        callback._registered_tensors = registered_tensors
+                    else:  # For now, we can use the old callback function. In the future we should improve this
+                        registered_tensors["loss"] = final_loss
+
+        def get_state(action: 'PtAction'):
+            """Helper function used to create a state for callbacks
+            """
+
+            class StateWrapper(dict):
+                def __init__(self, action):
+                    """A class that wraps a dictionary but adds the functions: restore_state_from and save_state_to
+                    which are helper functions for CheckpointCallback to use.
+                    The StateWrapper is a dictionary that contains the following mapping:
+                        "step" (int): the current training step
+                        "epoch" (int): the current epoch step
+                        "local_rank" (int): the local rank that the process is running on
+                        "global_rank" (int): the global rank that the process is running on
+                        "optimizers" (list): a list of optimizers defined during the training process
+                        "tensors" (TrainingState): A TrainingState object that can be used to access tensor values
+                    """
+                    self.action = action
+                    super().__init__(
+                        {
+                            "step": action.step,
+                            "tensors": action._training_state,
+                            "epoch": action.epoch,
+                            "local_rank": action.local_rank,
+                            "global_rank": action.global_rank,
+                            "optimizers": action.optimizers,
+                        }
+                    )
+
+                def restore_state_from(self, path):
+                    if os.path.isfile(path):
+                        # map_location could be cuda:<device_id> but cpu seems to be more
+                        # general since we are also saving step and epoch
+                        # load_state_dict should move the variables to the relevant device
+                        checkpoint = torch.load(path, map_location="cpu")
```

Comment:
Can you consider making it "backend-agnostic" using: nemo.backends.load
like in here:
https://github.com/NVIDIA/NeMo/blob/d91f349e46facb3f0943e2a4341557102a5bfd52/nemo/core/neural_graph.py#L1028

You know, one "agnostic" step at a time ;)